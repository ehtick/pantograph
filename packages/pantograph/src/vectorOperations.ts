import type { Vector } from "./definitions";

export const DEG2RAD = Math.PI / 180;

export const reprVector = (vector: Vector): string => {
  return `[${vector[0]}, ${vector[1]}]`;
};

export const sameVector = (
  [x0, y0]: Vector,
  [x1, y1]: Vector,
  precision = 1e-9
): boolean => {
  return Math.abs(x0 - x1) <= precision && Math.abs(y0 - y1) <= precision;
};

export const add = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 + x1, y0 + y1];
};

export const subtract = ([x0, y0]: Vector, [x1, y1]: Vector): Vector => {
  return [x0 - x1, y0 - y1];
};

export const squareLength = ([x, y]: Vector): number => {
  return x * x + y * y;
};

export const length = ([x, y]: Vector): number => {
  return Math.sqrt(squareLength([x, y]));
};

export const scalarMultiply = ([x0, y0]: Vector, scalar: number): Vector => {
  return [x0 * scalar, y0 * scalar];
};

export const squareDistance = (
  [x0, y0]: Vector,
  [x1, y1]: Vector = [0, 0]
): number => {
  return (x0 - x1) ** 2 + (y0 - y1) ** 2;
};

export const distance = (p0: Vector, p1: Vector = [0, 0]): number => {
  return Math.sqrt(squareDistance(p0, p1));
};

export function crossProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * y1 - y0 * x1;
}

export function dotProduct([x0, y0]: Vector, [x1, y1]: Vector): number {
  return x0 * x1 + y0 * y1;
}

export const angle = ([x0, y0]: Vector, [x1, y1]: Vector = [0, 0]): number => {
  return Math.atan2(y1 * x0 - y0 * x1, x0 * x1 + y0 * y1);
};

export function normalize([x0, y0]: Vector): Vector {
  const l = distance([x0, y0]);
  return [x0 / l, y0 / l];
}

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta) * r;
  const y = Math.sin(theta) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = distance([x, y]);
  const theta = Math.atan2(y, x);

  return [r, theta];
}

export function parallel(v1: Vector, v2: Vector, precision = 1e-9): boolean {
  const V1xV2 = crossProduct(v1, v2);

  const xLength = squareLength(v1);
  const yLength = squareLength(v2);

  return V1xV2 * V1xV2 < xLength * yLength * precision * precision;
}
