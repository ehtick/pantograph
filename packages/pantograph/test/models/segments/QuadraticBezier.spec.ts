import { describe, it, expect } from "vitest";

import { QuadraticBezier } from "../../../src/models/segments/QuadraticBezier.js";
import { add, scalarMultiply } from "../../../src/vectorOperations.js";
//import { debugImg, dpnt, drawBbox } from "../../debug.js";

describe("QuadraticBezier", () => {
  it("should detect when a point is on the curve", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0]);

    expect(curve.isOnSegment([0.75, 0.25])).toBe(true);
    expect(curve.isOnSegment([0, 0])).toBe(true);
    expect(curve.isOnSegment([1, 1])).toBe(true);
    expect(curve.translateX(5).isOnSegment([5.75, 0.25])).toBe(true);

    //debugImg([curve.translateY(5), dpnt([0.75, 5.25]), dpnt([0, 0])]);

    expect(curve.translateY(5).isOnSegment([0.75, 5.25])).toBe(true);

    expect(
      curve.rotate(-45, [0.75, 0.25]).translateX(5).isOnSegment([5.75, 0.25]),
    ).toBe(true);
  });

  it("handles point to param and param to point correctly", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1])
      .rotate(-15, [0.75, 0.25])
      .translateX(0.12);

    const intervals = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1];

    //debugImg([curve, ...intervals.map((i) => dpnt(curve.paramPoint(i)))]);

    intervals.forEach((i) => {
      expect(curve.pointToParam(curve.paramPoint(i))).toBeCloseTo(i);
    });
  });

  it("splits correctly in one point", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1])
      .translateX(0.12)
      .rotate(-15, [0.5, 0.5]);

    const curves = curve.splitAt([curve.paramPoint(0.66)]);

    expect(curves.length).toBe(2);

    //debugImg([{ shape: curve, color: "red" }, ...curves]);

    expect(curves[0].firstPoint).toEqual(curve.firstPoint);
    expect(curves[0].lastPoint).toEqual(curve.paramPoint(0.66));
    expect(curves[1].firstPoint).toEqual(curve.paramPoint(0.66));
    expect(curves[0].paramPoint(0.5)).toBeVector(curve.paramPoint(0.33));
    expect(curves[1].lastPoint).toBeVector(curve.lastPoint);
  });

  it("splits correctly in multiple points", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1])
      .translateX(0.12)
      .rotate(-15, [0.5, 0.5]);

    const curves = curve.splitAt([
      curve.paramPoint(0.75),
      curve.paramPoint(0.25),
    ]);

    //debugImg([{ shape: curve, color: "red" }, ...curves]);

    expect(curves.length).toBe(3);

    expect(curves[0].firstPoint).toEqual(curve.firstPoint);
    expect(curves[0].lastPoint).toEqual(curve.paramPoint(0.25));
    expect(curves[1].firstPoint).toEqual(curve.paramPoint(0.25));
    expect(curves[1].paramPoint(0.5)).toBeVector(curve.paramPoint(0.5));
    expect(curves[1].lastPoint).toEqual(curve.paramPoint(0.75));
    expect(curves[2].firstPoint).toEqual(curve.paramPoint(0.75));
    expect(curves[2].lastPoint).toEqual(curve.lastPoint);
  });

  it("splits correctly in multiple points and ignores duplicates", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1])
      .translateX(0.12)
      .rotate(-15, [0.5, 0.5]);

    const curves = curve.splitAt([
      curve.paramPoint(0.75),
      curve.paramPoint(1),
      curve.paramPoint(0.25),
      curve.paramPoint(0),
      curve.paramPoint(0.7500000001),
    ]);

    //debugImg([{ shape: curve, color: "red" }, ...curves]);

    expect(curves.length).toBe(3);

    expect(curves[0].firstPoint).toEqual(curve.firstPoint);
    expect(curves[0].lastPoint).toEqual(curve.paramPoint(0.25));
    expect(curves[1].firstPoint).toEqual(curve.paramPoint(0.25));
    expect(curves[1].paramPoint(0.5)).toBeVector(curve.paramPoint(0.5));
    expect(curves[1].lastPoint).toEqual(curve.paramPoint(0.75));
    expect(curves[2].firstPoint).toEqual(curve.paramPoint(0.75));
    expect(curves[2].lastPoint).toEqual(curve.lastPoint);
  });

  it("splits correctly at many points", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1])
      .rotate(-75, [0.5, 0.5])
      .translateX(0.12);

    const curves = curve.splitAt([
      curve.paramPoint(0.1),
      curve.paramPoint(0.2),
      curve.paramPoint(0.3),
      curve.paramPoint(0.4),
      curve.paramPoint(0.5),
      curve.paramPoint(0.6),
      curve.paramPoint(0.75),
      curve.paramPoint(0.8),
      curve.paramPoint(0.25),
      curve.paramPoint(0.9),
    ]);

    //debugImg([{ shape: curve, color: "red" }, ...curves]);

    expect(curves.length).toBe(11);
    expect(curves).toMatchSnapshot();
  });

  it("computes the distance to a point close to the curve correctly", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [0.1, 1.1]);

    const distantPoint = (param: number, distance: number) => {
      const p0 = curve.paramPoint(param);
      const p1 = curve.normalAt(p0);
      return add(scalarMultiply(p1, distance), p0);
    };

    expect(curve.distanceFrom(distantPoint(0, 0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.2, 0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.4, 0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.6, 0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.8, 1))).toBeCloseTo(1);
    expect(curve.distanceFrom(distantPoint(0.9, 0.1))).toBeCloseTo(0.1);

    expect(curve.distanceFrom(distantPoint(0.1, -0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.2, -0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.4, -0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.6, -0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.8, -0.1))).toBeCloseTo(0.1);
    expect(curve.distanceFrom(distantPoint(0.9, -0.1))).toBeCloseTo(0.1);
  });

  it("computes the distance to a point far from the endpoint curve correctly", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]);

    //debugImg([{ shape: curve, color: "red" }, dpnt([-2, 0]), dpnt([2, 1])]);

    expect(curve.distanceFrom([-2, 0])).toBeCloseTo(2);
    expect(curve.distanceFrom([2, 1])).toBeCloseTo(1);
  });

  it("compute the bounding box correctly", () => {
    const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]);
    //debugImg([{ shape: curve, color: "red" }, drawBbox(curve)]);
    expect(curve.boundingBox).toMatchSnapshot();

    const curve2 = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]).rotate(15);
    //debugImg([{ shape: curve2, color: "red" }, drawBbox(curve2)]);
    expect(curve2.boundingBox).toMatchSnapshot();

    const curve3 = new QuadraticBezier([0, 0], [1, 1], [1, 0.5]).rotate(45);
    //debugImg([{ shape: curve3, color: "red" }, drawBbox(curve3)]);
    expect(curve3.boundingBox).toMatchSnapshot();

    const curve4 = new QuadraticBezier([-0.5, 0], [1, 0], [-1, 1]);
    //debugImg([{ shape: curve4, color: "red" }, drawBbox(curve4)]);
    expect(curve4.boundingBox).toMatchSnapshot();
  });
});
