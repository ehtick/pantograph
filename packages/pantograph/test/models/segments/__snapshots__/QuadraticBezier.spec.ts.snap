// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuadraticBezier > compute the bounding box correctly 1`] = `
BoundingBox {
  "xMax": 1.000000001,
  "xMin": -1e-9,
  "yMax": 1.000000001,
  "yMin": -1e-9,
}
`;

exports[`QuadraticBezier > compute the bounding box correctly 2`] = `
BoundingBox {
  "xMax": 0.7533647425304433,
  "xMin": -1e-9,
  "yMax": 1.224744872391589,
  "yMin": -1e-9,
}
`;

exports[`QuadraticBezier > compute the bounding box correctly 3`] = `
BoundingBox {
  "xMax": 0.17677669629663695,
  "xMin": -1e-9,
  "yMax": 1.414213563373095,
  "yMin": -1e-9,
}
`;

exports[`QuadraticBezier > compute the bounding box correctly 4`] = `
BoundingBox {
  "xMax": 1.000000001,
  "xMin": -0.6000000010000001,
  "yMax": 0.500000001,
  "yMin": -1e-9,
}
`;

exports[`QuadraticBezier > splits correctly at many points 1`] = `
[
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.04316872707734828,
      0.7595489984153921,
    ],
    "firstPoint": [
      0.007627564304205525,
      0.8535533905932737,
    ],
    "lastPoint": [
      0.08384910600977839,
      0.6772744168612214,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.12452948494220849,
      0.5949998353070506,
    ],
    "firstPoint": [
      0.08384910600977839,
      0.6772744168612214,
    ],
    "lastPoint": [
      0.17034908003392593,
      0.5244550643765906,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.19325887757978466,
      0.48918267891136064,
    ],
    "firstPoint": [
      0.17034908003392593,
      0.5244550643765906,
    ],
    "lastPoint": [
      0.21745347916546517,
      0.45684274610205844,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.24164808075114574,
      0.4245028132927562,
    ],
    "firstPoint": [
      0.21745347916546517,
      0.45684274610205844,
    ],
    "lastPoint": [
      0.26712748637664807,
      0.39509533313938155,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.31808629762765284,
      0.3362803728326325,
    ],
    "firstPoint": [
      0.26712748637664807,
      0.39509533313938155,
    ],
    "lastPoint": [
      0.374184325037945,
      0.28919522314959434,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.4302823524482371,
      0.24211007346655614,
    ],
    "firstPoint": [
      0.374184325037945,
      0.28919522314959434,
    ],
    "lastPoint": [
      0.49151959601781653,
      0.20675473440722872,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.5527568395873961,
      0.17139939534790127,
    ],
    "firstPoint": [
      0.49151959601781653,
      0.20675473440722872,
    ],
    "lastPoint": [
      0.6191332993162628,
      0.14777386691228483,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.7186979889095635,
      0.1123355742588599,
    ],
    "firstPoint": [
      0.6191332993162628,
      0.14777386691228483,
    ],
    "lastPoint": [
      0.8298259148612596,
      0.10328935550878462,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.8668685568451655,
      0.10027394925875896,
    ],
    "firstPoint": [
      0.8298259148612596,
      0.10328935550878462,
    ],
    "lastPoint": [
      0.9051960028688792,
      0.10019099566466208,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      0.981850894916321,
      0.1000250884764684,
    ],
    "firstPoint": [
      0.9051960028688792,
      0.10019099566466208,
    ],
    "lastPoint": [
      1.0636450031230495,
      0.11158899191198328,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
  QuadraticBezier {
    "_boundingBox": null,
    "_polynomialCoefficients": null,
    "controlPoint": [
      1.145439111329778,
      0.12315289534749924,
    ],
    "firstPoint": [
      1.0636450031230495,
      0.11158899191198328,
    ],
    "lastPoint": [
      1.2323724356957944,
      0.14644660940672616,
    ],
    "precision": 1e-9,
    "segmentType": "QUADRATIC_BEZIER",
  },
]
`;
